internal/oauth2/lazy_token.go

```diff
@@ -0,0 +1,118 @@
+package oauth2
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Tokener is an interface that defines the Token method.
+type Tokener interface {
+	Token(context.Context) string
+}
+
+// LazyToken is a struct that handles token acquisition and caching.
+type LazyToken struct {
+	url          string
+	clientId     string
+	clientSecret string
+	token        string
+	offlineToken string
+	expiration   time.Time
+	mutex        sync.Mutex
+}
+
+type tokenResponse struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int    `json:"expires_in"`
+	TokenType   string `json:"token_type"`
+}
+
+// NewLazyToken creates a new LazyToken instance.
+func NewLazyToken(url, clientId, clientSecret, offlineToken string) *LazyToken {
+	return &LazyToken{
+		url:          url,
+		clientId:     clientId,
+		clientSecret: clientSecret,
+		offlineToken: offlineToken,
+		token:        "",
+		expiration:   time.Now(),
+	}
+}
+
+// Token returns a valid token, acquiring a new one if necessary.
+func (lt *LazyToken) Token(ctx context.Context) string {
+	lt.mutex.Lock()
+	defer lt.mutex.Unlock()
+
+	if lt.token == "" || lt.expiration.Before(time.Now().Add(time.Minute)) {
```

Comment:
The time condition is I believe not correct. Token should be refreshed if:

* Token hasn’t been fetched yet (is empty)
* Current time is older than expiration time of already fetched token minus one minute.

Write test for both use cases.